### PR TITLE
Compile with C++14

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
C++14 adds very useful things for embedded, especially in terms of constexpr.

 #### Summary of Changes
Switch C++ standard from 11 to 14 for CHIP

### CODE REVIEW

File: docs/style/coding/CODING_STYLE_GUIDE.adoc [hunk 1]
```diff
@@ -151,50 +151,12 @@ leveraged through toolchain-compatibility preprocessor macros.
...
-#else
-#define __chipFINAL
-#endif
-----
-[.text-center]
-*Listing 2.1.* Using the C preprocessor to provide a portability mnemonic
-for the C{plusplus}11 and later final keyword.
+C{plusplus}14 is considered pervasive enough to be used. As compilers start 
+supporting standards suche as C{plusplus}17, C{plusplus}20 and beyond,
+CHIP may follow suite.
```
Comment (gerickson) [inline docs/style/coding/CODING_STYLE_GUIDE.adoc:159]: suite -> suit.

File: src/lib/support/CHIPArgParser.cpp [hunk 2]
```diff
@@ -43,6 +43,7 @@
 #include <stdarg.h>
 #include <stdint.h>
 #include <string.h>
+#include <strings.h>
```
Comment (Damian-Nordic) [inline src/lib/support/CHIPArgParser.cpp:46]: I think this shouldn't be needed once you switched to gnu++14

**Review by Damian-Nordic — APPROVED** [review]

File: src/lib/support/CHIPArgParser.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/support/CHIPArgParser.cpp:46]: It was indeed not needed, however purely from a `man strcasecmp` it seems to be technically correct.

**Review by jelderton — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
Looks great, if the platform support is there....

**Review by mspang — APPROVED** [review]

**Pull request merged**